### ISSUE TITLE
[css-ui] Readonly form control and user-select

### ISSUE THREAD

**Issue by upsuper** [w3c/csswg-drafts#285]
Currently, for user-select, CSS UI 4 says
> on editable elements where the computed value is always `contain` regardless of the specified value

and it defines "editable elements" as
> an editable element is either an editing host or a mutable form control with textual content, such as textarea.

It seems this would exclude readonly form controls from the computation rule above, because according to the HTML Standard, readonly form controls are not mutable. But I think that rule should be applied to those controls as well.

**Comment by frivoal** [comment]
This suggestion does seem to match the current behavior of UAs. If you create a `<textarea readonly>` element, the selection will be trapped in it.

Do you have a suggestion for the phrasing? The following would probably work, but it looks a bit clunky:

> an editable element is either an editing host or a form control with normally mutable textual content such as textarea, even if this form control is in non mutable state.

"normally mutable but in non mutable state" is an attempt to include `<textarea readonly>` but not `<label>`.

**Comment by upsuper** [comment]
Actually, UAs other than Gecko allow text in disabled form controls to be selected as well... And there is [a bug](https://bugzilla.mozilla.org/show_bug.cgi?id=195361) which urges Gecko to follow that behavior... But I don't think that makes much sense so I opened [an issue](https://bugs.chromium.org/p/chromium/issues/detail?id=626581) for Chromium, hoping they could change their behavior.

For phrasing, I don't have any idea... Otherwise I would have included that in my post :)

I wonder whether it makes more sense to use user agent stylesheet to do that, rather than making it a computation rule.

**Comment by frivoal** [comment]
> Actually, UAs other than Gecko allow text in disabled form controls to be selected as well...

Yes, it can be selected, but the selection cannot escape the element.

The fact that this is true no only with readonly but also with disabled is why I used the "non mutable state" wording in the proposed phrasing, instead of specifically referring to the readonly attribute.

As for whether that should be done via computation or via UA stylesheet, that's a good question, but regardless of the answer, we should put the desired behavior in the spec.

**Comment by upsuper** [comment]
> Yes, it can be selected, but the selection cannot escape the element.

I meant, I don't think disabled form controls should be selectable...

**Comment by frivoal** [comment]
Actually, I don't think we want to do this with the UA stylesheet, as that would mean that authors could override it, which may not be implementable, since form controls are often rendered with native elements, and having a selection extent from one to outside may not be doable.

So I suggestn going with the phrasing proposed in https://github.com/w3c/csswg-drafts/issues/285#issuecomment-231321757

I've checked if we could simply say that it computes to contain on replaced elements in general, but that does not seem to be true, as for instance there's not contain effect on embedded svg.

**Comment by frlinw** [comment]
Even if user-select default value is `contain` for readonly input text/textarea, why the user couldn't force it to `none` ?

**Comment by frivoal** [comment]
@frlinw on non mutable ones, that seems reasonable. On actually mutable ones, it's not clear how editing would work, so none should not be allowed.

**Comment by frivoal** [comment]
Edits will be made in a pull request first, to give a chance to @upsuper to check.

**Comment by frivoal** [comment]
I made https://github.com/w3c/csswg-drafts/pull/2961 to attempt to close this. Agenda+
